pkg/scheduler/queue/querier_worker_group_queue_algorithm.go

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+type querierWorkerPrioritizationQueueAlgo struct {
+	currentQuerierWorker  int32
+	currentNodeOrderIndex int
+	nodeOrder             []string
+	nodesChecked          int
+}
+
+func (qa *querierWorkerPrioritizationQueueAlgo) SetCurrentQuerierWorker(workerID int32) {
+	qa.currentQuerierWorker = workerID
+	if len(qa.nodeOrder) == 0 {
+		qa.currentNodeOrderIndex = 0
+	} else {
+		qa.currentNodeOrderIndex = int(workerID) % len(qa.nodeOrder)
+	}
+}
+
+func (qa *querierWorkerPrioritizationQueueAlgo) incrementWrapCurrentNodeOrderIndex() {
+	qa.currentNodeOrderIndex++
+	if qa.currentNodeOrderIndex >= len(qa.nodeOrder) {
+		qa.currentNodeOrderIndex = 0
+	}
+}
+
+func (qa *querierWorkerPrioritizationQueueAlgo) checkedAllNodes() bool {
+	return qa.nodesChecked == len(qa.nodeOrder)
+}
+
+func (qa *querierWorkerPrioritizationQueueAlgo) addChildNode(parent, child *Node) {
+	// add childNode to n.queueMap
+	parent.queueMap[child.Name()] = child
+
+	if qa.currentNodeOrderIndex == localQueueIndex {
+		// special case; cannot slice into nodeOrder with index -1
+		// place at end of slice, which is the last slot before the local queue slot
+		qa.nodeOrder = append(qa.nodeOrder, child.Name())
+	} else if qa.currentNodeOrderIndex == 0 {
+		// special case; since we are at the beginning of the order,
+		// only a simple append is needed to add the new node to the end,
+		// which also creates a more intuitive initial order for tests
+		qa.nodeOrder = append(qa.nodeOrder, child.Name())
+	} else {
+		// insert into the order behind current child queue index
+		// to prevent the possibility of new nodes continually jumping the line
+		qa.nodeOrder = append(
+			qa.nodeOrder[:qa.currentNodeOrderIndex],
+			append(
+				[]string{child.Name()},
+				qa.nodeOrder[qa.currentNodeOrderIndex:]...,
+			)...,
+		)
```

Comment:
```suggestion
		qa.nodeOrder = slices.Insert(qa.nodeOrder, qa.currentNodeOrderIndex, child.Name())
```

Reply:
The same patterns exist elsewhere too

Reply:
whew thanks I didn't know those existed, the nested appends are so hard to look at

Reply:
(resolving not because this isn't helpful but to keep the comment view focused on the benchmark results)